### ISSUE TITLE
[Yul] sub-objects are not accessible

### ISSUE THREAD

**Issue by chriseth** [ethereum/solidity#6739]
In yul, sub-objects of objects should be accessible via `.`, but they are not.

Related: https://github.com/ethereum/solidity/issues/6514

**Comment by axic** [comment]
If I recall correctly, this is the example here:
```
object "A" {
  code {
    pop(dataoffset("A")) // works
    pop(dataoffset("B")) // works
    pop(dataoffset("A.B")) // fails, but should work
  }
  
  object "B" {
    code {
      invalid()
    }
  }
}
```

```
Error: Unknown data object "A.B".
 --> yulobj.sol:5:9:
  |
5 |     pop(dataoffset("A.B"))
  |         ^^^^^^^^^^
```

**Comment by ekpyron** [comment]
This will be needed to implement ``type(C).runtimeCode`` (see https://github.com/ethereum/solidity/pull/8867#discussion_r421843356).

**Comment by mijovic** [comment]
Can be solved by adding names like `objectName.subObject` to `dataNames` for Object in `libYul/Object.cpp`
Not sure if this would be proper solution?

**Comment by mijovic** [comment]
> Can be solved by adding names like `objectName.subObject` to `dataNames` for Object in `libYul/Object.cpp`
> Not sure if this would be proper solution?

Or better option to parse and search for subobjects in `AsmAnalysis.cpp`

**Comment by ekpyron** [comment]
Both adding ``objectName.subObject`` to ``dataNames`` (i.e. to the "global scope" IIUC) and descending into objects when searching sounds viable.
But note that this needs to be recursive (at least I'd make it so), i.e. ``objectName.subObject.subSubObject`` has to work as well - and we need to make sure there are no conflicts, e.g. the following currently compiles:
```
object "obj" {
        code {
        }
        object "subobj" {
                code {
                }
                object "subsubobj" {
                        code {
                        }
                }
        }
        object "subobj.subsubobj" {
                code {
                }
        }
}
```
Maybe dots ``.`` should not be allowed in names of yul objects (ping @chriseth)? Or should ``subobj.subsubobj`` just conflict with the inner ``subsubobj`` of the other ``subobj``?

**Comment by chriseth** [comment]
It makes sense to disallow dots in yul object names.

By the way, do we know if this will work in the end? Can we reach deep into Assembly objects like that?

**Comment by mijovic** [comment]
> It makes sense to disallow dots in yul object names.
> 
> By the way, do we know if this will work in the end? Can we reach deep into Assembly objects like that?

We can't go deep into Assembly objects, it just can take direct subobjects.

**Comment by mijovic** [comment]
> If I recall correctly, this is the example here:
> 
> ```
> object "A" {
>   code {
>     pop(dataoffset("A")) // works
>     pop(dataoffset("B")) // works
>     pop(dataoffset("A.B")) // fails, but should work
>   }
>   
>   object "B" {
>     code {
>       invalid()
>     }
>   }
> }
> ```
> 
> ```
> Error: Unknown data object "A.B".
>  --> yulobj.sol:5:9:
>   |
> 5 |     pop(dataoffset("A.B"))
>   |         ^^^^^^^^^^
> ```

@axic Does it makes sense to implement `A.B` at all in cases like this. I think it is not needed, and we need to implement `B.C` (if B had subobject C) like done in #8967 in test `subObjectAccess.yul`?

**Comment by axic** [comment]
@mijovic Sorry missed the question. Do you mean that the path should always be relative to the current object
a) and not be an absolute path
b) not include the object itself

I agree with a), but not sure about b).

@ekpyron @chriseth any opinions?

**Comment by chriseth** [comment]
What is easier for the code generator? Having to use `A.` all the time seems a bit redundant.

**Comment by mijovic** [comment]
I would allow from user side to do `A.`, but internally would use it as relative path without object itself, in this case `A.`

**Comment by bshastry** [comment]
Is this fixed by https://github.com/ethereum/solidity/pull/9235 ?

**Comment by mijovic** [comment]
> Is this fixed by #9235 ?

Yes it is fixed by #9235
